### ISSUE TITLE
Support Rails-y HTML forms

### ISSUE THREAD

**Issue by maxfurman** [tiangolo/fastapi#614]
**Is your feature request related to a problem? Please describe.**
I'm trying to point some existing forms to a FastAPI backend. They contain properties of a model like `<input type="text" name="user[first_name]">`. I would like to use a Pydantic model that describes the entire User object, instead of putting each form param as its own argument to the handler, or reading and parsing the entire raw request body.

**Describe the solution you'd like**
```
class User(BaseModel):
    first_name: str
    ... more properties ...

@app.post('/users/')
async def create_user(user: User = Form(...)):
    do_something(user)
```

**Describe alternatives you've considered**
Being able to grab the entire posted form body as a Dict would work too, I think

**Comment by dmontagu** [comment]
For what it's worth, you can do something similar to using a pydantic model now (as opposed to listing things out in the endpoint signature), you just can't do it using a pydantic model specifically.


```python
from fastapi import FastAPI, Depends
from starlette.testclient import TestClient

class UserForm:
    def __init__(self, first_name: str, other: str):
        self.first_name = first_name
        self.other = other

app = FastAPI()


@app.get("/")
def create_user(user_form: UserForm = Depends()):
    return [user_form.first_name, user_form.other]


client = TestClient(app)
print(client.get("/").json())
# {
#     "detail": [
#         {
#             "loc": ["query", "first_name"],
#             "msg": "field required",
#             "type": "value_error.missing",
#         },
#         {
#             "loc": ["query", "other"],
#             "msg": "field required",
#             "type": "value_error.missing",
#         },
#     ]
# }
print(client.get("/", params={"first_name": "hello", "other": "world"}).json())
# ['hello', 'world']
```

If the only thing you don't like is that you have to write out an init signature, at least in python3.7 you can use `dataclasses` too:
```python
from dataclasses import dataclass

@dataclass
class UserForm:
    first_name: str
    other: str
```
(if you swap out `UserForm` in the longer example above with this definition, the results are exactly the same)

**Note that you'll still get pydantic-based validation of the values since fastapi is already doing that when it obtains the dependencies for use in the dependency signatures.**

-----

If you specifically want the ability to use pydantic models for this purpose, that might be a reasonable feature request; as far as I am aware that is not currently possible. (Although I believe @tiangolo has pushed back against this in other issues in the past.)

But I figured there's a chance the above approach addresses your workflow.

**Comment by dmontagu** [comment]
Also, I realize you wanted to use `Form` rather than query parameters; that works just as well:

```python
class UserForm:
    def __init__(self, first_name: str = Form(...), other: str = Form(...)):
        self.first_name = first_name
        self.other = other

# OR

@dataclass
class UserForm:
    first_name: str = Form(...)
    other: str = Form(...)

...

print(client.get("/", data={"first_name": "hello", "other": "world"}).json())
```

Youl'll get the same output as above

**Comment by maxfurman** [comment]
Thanks! That's much better than grabbing each attribute as an argument. Is there a way to untangle nested form inputs? For example, to turn `user[first_name]=Rick&user[last_name]=Sanchez` formdata into `{'user': {'first_name': 'Rick', 'last_name': 'Sanchez'}}`, and then coerce that into a User instance?

**Comment by dmontagu** [comment]
You’d probably want to look into the python multipart form dependency to see how it parses the data. I’m not sure how it works; it wouldn’t surprise me if this were possible, but it also wouldn’t surprise me if it weren’t. (You might also want/need to look at how the starlette Request class handles form data.)

----

In general, I would recommend using json instead of form data once dictionaries come into play — I’ve just noticed a lot of inconsistency in how form data like this is serialized/parsed, and it seems disproportionately likely that you could run into a painful client and/or server issue down the line due to the atypical form data usage (despite what any RFCs may or may not say).

Just my 2c though.

**Comment by maxfurman** [comment]
I haven't had any luck finding a Python library or builtin to parse HTML FormData formatted[like][this] - which I find strange since that pattern is pretty common In The Wild™. Maybe I'm missing something. I'm trying to replace some of my backend routes without rewriting any of the frontend forms, but that may not be possible.

**Comment by tiangolo** [comment]
Yep, as @dmontagu says. I imagine that's probably a Rails convention or similar? I haven't seen it.

It's not part of any standard, and if I remember correctly is not even possible to declare it in OpenAPI. The problem is that, as it's not in any standard, there could be lots of similar custom encodings.

That would require a custom parser for those forms, but is not something that something that I have seen before, so I don't know of any package that could do it 😔

If you find one, or implement it yourself, the way to tie it together would be right as @dmontagu says. Just that you would parse those custom form names/keys (`str`) to extract their nested data. A similar format I can think of is TOML, so maybe you can use parts of it. But it's not completely the same.
